### PR TITLE
Docs: Implement Controls block

### PR DESCRIPTION
**Pull request by tmeasday**
Telescoping on https://github.com/storybookjs/storybook/pull/20664

Issue: N/A

## What I did

- Very similar to #20664, implemented a controls block. There is some small duplication with the deprecated `ArgTable` component, but it is very simple.

- Use it in DocsPage

## How to test

- See stories
- Run an Autodocs page in a sandbox.

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [x] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/ui/blocks/src/blocks/Controls.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+// NOTE: this will throw with no of prop
+export const OfStoryUnattached: Story = {
+  parameters: { attached: false },
+  args: {
+    of: ExampleStories.NoParameters,
+  },
+};
+
+export const IncludeProp: Story = {
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Controls.stories.tsx:32]: do you think it's overkill to add stories that has both props and parameters to ensure props take priority?

File: code/ui/blocks/src/blocks/Controls.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+  const [args, setArgs] = useState(storyContext.args);
+  useEffect(() => {
+    const cb = (changed: { storyId: string; args: Args }) => {
+      if (changed.storyId === storyId) {
+        setArgs(changed.args);
+      }
+    };
+    context.channel.on(STORY_ARGS_UPDATED, cb);
+    return () => context.channel.off(STORY_ARGS_UPDATED, cb);
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Controls.tsx:44]:
```suggestion
    const onArgsUpdated = (changed: { storyId: string; args: Args }) => {
      if (changed.storyId === storyId) {
        setArgs(changed.args);
      }
    };
    context.channel.on(STORY_ARGS_UPDATED, onArgsUpdated);
    return () => context.channel.off(STORY_ARGS_UPDATED, onArgsUpdated);
```

File: code/ui/blocks/src/blocks/Controls.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  }, [storyId, context.channel]);
+  const updateArgs = useCallback(
+    (updatedArgs) => context.channel.emit(UPDATE_STORY_ARGS, { storyId, updatedArgs }),
+    [storyId, context.channel]
+  );
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Controls.tsx:49]: it doesn't create any problems with closures or anything, that `updateArgs` is defined twice here?

File: code/ui/blocks/src/blocks/Controls.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  const storyContext = context.getStoryContext(story);
+
+  const [globals, setGlobals] = useState(storyContext.globals);
+  useEffect(() => {
+    const cb = (changed: { globals: Globals }) => {
+      setGlobals(changed.globals);
+    };
+    context.channel.on(GLOBALS_UPDATED, cb);
+    return () => context.channel.off(GLOBALS_UPDATED, cb);
+  }, [context.channel]);
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Controls.tsx:67]:
```suggestion
  useEffect(() => {
    const onGlobalsUpdated = (changed: { globals: Globals }) => {
      setGlobals(changed.globals);
    };
    context.channel.on(GLOBALS_UPDATED, onGlobalsUpdated);
    return () => context.channel.off(GLOBALS_UPDATED, onGlobalsUpdated);
  }, [context.channel]);
```

File: code/ui/blocks/src/examples/ControlsParameters.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import type { Meta, StoryObj } from '@storybook/react';
+import { ControlsParameters } from './ControlsParameters';
+
+/**
+ * Reference stories to be used by the Controls stories
+ */
+const meta = {
+  title: 'Example/Stories for Controls',
```
Comment (JReinhold) [inline code/ui/blocks/src/examples/ControlsParameters.stories.tsx:8]: we have a whole section called Controls, so we might want to make this more explicit.
```suggestion
  title: 'Example/Stories for the Controls Block',
```

File: code/ui/blocks/src/examples/ControlsParameters.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+} satisfies Meta<typeof ControlsParameters>;
+
+export default meta;
+type Story = StoryObj<typeof meta>;
+
+/**
+ * This is the primary mode for the button
+ *
+ * _this description was written as a comment above the story_
+ */
```
Comment (JReinhold) [inline code/ui/blocks/src/examples/ControlsParameters.stories.tsx:29]:
```suggestion
```

**Review by JReinhold — CHANGES_REQUESTED** [review]
[code/ui/blocks/src/argsHelpers.ts](https://github.com/storybookjs/storybook/pull/20683/files#diff-08e29b0c9949afcacee51a8d3d9063342471dec592f6f7971e08b83718107950) is an empty file?

Looks very clean!

File: code/ui/blocks/src/blocks/Controls.stories.tsx (hunk 1 above)
Comment (tmeasday) [reply code/ui/blocks/src/blocks/Controls.stories.tsx:32]: I'm not sure. We might have to do a lot of them, theoretically. WDTY?

File: code/ui/blocks/src/blocks/Controls.tsx (hunk 3 above)
Comment (tmeasday) [reply code/ui/blocks/src/blocks/Controls.tsx:49]: Twice? Do you mean once in `useArgs` and once in the `Controls` block?

**Review by tmeasday — COMMENTED** [review]
Apply @JReinhold's suggestions

File: code/ui/blocks/src/blocks/Controls.tsx (hunk 3 above)
Comment (JReinhold) [reply code/ui/blocks/src/blocks/Controls.tsx:49]: no sorry, I read both as `updateArgs` but line 47 has an extra "d". never mind. 🤦

File: code/ui/blocks/src/blocks/Controls.stories.tsx (hunk 1 above)
Comment (tmeasday) [reply code/ui/blocks/src/blocks/Controls.stories.tsx:32]: Created a ticket for this.

**Comment by Alishfie** [comment]
> Telescoping on https://github.com/storybookjs/storybook/pull/20664
> 
> Issue: N/A
> 
> ## What I did
> 
> - Very similar to #20664, implemented a controls block. There is some small duplication with the deprecated `ArgTable` component, but it is very simple.
> 
> - Use it in DocsPage
> 
> ## How to test
> 
> - See stories
> - Run an Autodocs page in a sandbox.
> 
> 
> ## Checklist
> 
> 
> 
> - [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
> - [x] Make sure to add/update documentation regarding your changes
> - [x] If you are deprecating/removing a feature, make sure to update
>       [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)
> 
> #### Maintainers
> 
> - [x] If this PR should be tested against many or all sandboxes,
>       make sure to add the `ci:merged` or `ci:daily` GH label to it.
> - [x] Make sure this PR contains **one** of the labels below.
> 
> `["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
> 
> 
>

**Review by JReinhold — APPROVED** [review]

**Pull request merged**